Benchmark and perf improvement of Pivot.Get

mpizenberg opened the pull request:
Hi, I was using elm-pivot last year but eventually made my own smaller local zipper for mainly two reasons:

* to reduce bundle size (doesn't make sense anymore with DCE in 0.19)
* because the `hasL` function had a big perf cost (due to use of `getL`)

I figured I could contribute back with some small internal change of `Pivot.Get` functions. I added benchmarks to make sure I wasn't blindly worsening them. In firefox, I'm getting the results depicted in the following image. Let me know what you think.

![pivot-benchmark](https://user-images.githubusercontent.com/2905865/46403894-4f748d80-c736-11e8-8732-d14985bf8540.png)

yotamDvir commented:
Thanks! This is great.

Using `getL` in `hasL` makes little sense. I also like the optimization of `getA`, even though it's less significant. The `length*` functions could be similarly optimized.

In terms of package structuring, I think `reversePrepend` should be moved to `Pivot.Utilities`. Of course, there is no need to keep the unoptimized versions of functions, I'm assuming you did so just to get the benchmarks.

Looking through the library I got an idea on how to optimize `sortWith` (the left side could be compared using an inverted comparing function), but that may be better done separately.

One thing is bothering me: why does `getR` in `hasR` have such a performance impact? If you know, do tell!

mpizenberg commented:
> In terms of package structuring, I think reversePrepend should be moved to Pivot.Utilities.

Agreed, I prefered to keep changes locally before you have a look. I'm reaching the end of the day here, feel free to make changes that match your code organization, you have write access to this branch with PR setup.

> Of course, there is no need to keep the unoptimized versions of functions, I'm assuming you did so just to get the benchmarks.

Yep, just for validation before actually changing the original versions. After change, we'd have to remove half the benchmarks but could keep the other around I guess.

> One thing is bothering me: why does getR in hasR have such a performance impact? If you know, do tell!

It boils down to elm code generation actually. Using composition generates more function calls, which I guess put more work on the browser optimizations. Here are both versions generated by the elm compiler (not with --optimize flag, didn't check).

```js
// hasR
var author$project$Pivot$Get$hasR = A2(
	elm$core$Basics$composeR,
	author$project$Pivot$Get$getR,
	A2(elm$core$Basics$composeR, elm$core$List$isEmpty, elm$core$Basics$not));

// hasROptimized
var author$project$Pivot$Get$hasROptimized = function (_n0) {
	var _n1 = _n0.b;
	var r = _n1.b;
	return !elm$core$List$isEmpty(r);
};
```

mpizenberg commented:
Ok I've moved `reversePrepend` to `Pivot.Utilities` and renamed it `reversePrependList` to follow the pattern I saw with `assertList`.

Regarding the `hasR` question. I just ran the benchmark with chromium. I got similar results for all tests except `hasR` which performed equivalently with `hasROptimized`. I guess the chromium JIT compiler is more powerful than firefox one and recognizes the simplification in this case.

If the code is ok for you, don't hesitate to delete and replace the functions.

PS: I've carefully formatted "by hand" to try follow elm-format(@0.7.0-exp) which I guess is the one used here. I suppose you didn't do the switch to 0.8 because of `(..)` being expanded?

yotamDvir commented:
![screen shot 2018-10-06 at 9 35 31](https://user-images.githubusercontent.com/14016623/46568856-ba9bab00-c954-11e8-845a-01970a5af133.png)

yotamDvir commented:
> PS: I've carefully formatted "by hand" to try follow elm-format(@0.7.0-exp) which I guess is the one used here. I suppose you didn't do the switch to 0.8 because of (..) being expanded?

Ah, I hope that wasn't too much work. Just didn't know there was an update until you mentioned it.

mpizenberg commented:
I guess these are chromium benchmarks since lengthR performs the same with optimized.

> Ah, I hope that wasn't too much work.

No it was fine ^^. Actually the 0.8 version came out a little bit after the official version of 0.19. So if you updated your lib in the preparation period you would not have noticed it. I got caught also ^^.

yotamDvir commented:
> Regarding the hasR question. I just ran the benchmark with chromium. I got similar results for all tests except hasR which performed equivalently with hasROptimized. I guess the chromium JIT compiler is more powerful than firefox one and recognizes the simplification in this case.

Thanks for the analysis by the way. Makes me wonder how much can be optimized here by simply using less composition. Would become ugly if overdone though...

> I guess these are chromium benchmarks since lengthR performs the same with optimized.

Nope, Firefox on a Mac. But I'd like to keep it if only to be consistent with the other `length*` implementations.

mpizenberg commented:
> Makes me wonder how much can be optimized here by simply using less composition. Would become ugly if overdone though...

Similarly, out of curiosity, I just benchmarked `removeGoL` and `removeGoR` (which uses `mirrorM`) and turns out the mirrored version is 4x slower in my Firefox, and 2x slower in my Chromium, so there is definitely a cost in using abstractions.

yotamDvir commented:
That's a bit disappointing. I'll open an issue.
